sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainer.java

```diff
@@ -0,0 +1,201 @@
+ * metric updates produced from the executing code by interactions with the user-facing metric
+ * interfaces (eg., {@link Counter}).
+ * *
+ * <p>For a given runner to support metrics it is currently necessary to do 3 things:
+ * <ol>
+ *   <li>Create a {@link MetricsContainer} for each scope that metrics will be reported at.
+ *   </li>
+ *   <li>Make sure to call {@link #setMetricsContainer}
+ *   </li>
+ *   <li>Use {@link #getCumulative()} to get {@link org.apache.beam.sdk.metrics.MetricUpdates}
+ *   representing all the metric changes and report/aggregate those appropriately.
+ *   </li>
+ *   <li>Optionally, use {@link #getDeltas()} to get only the changes to a metric since the last
+ *   time and report those periodically. After reporting deltas and before the next call to
+ *   {@link #getDeltas()} the runner should invoke {@link #commitDeltas} to indicate that those
+ *   deltas have been incorporated.
+ *   </li>
+ * </ol>
+ */
+@Experimental(Kind.METRICS)
+public class MetricsContainer {
+
+  private static final ThreadLocal<MetricsContainer> CONTAINER_FOR_THREAD =
+      new ThreadLocal<MetricsContainer>();
+  private final AtomicBoolean committedFirstDelta = new AtomicBoolean(false);
+
+  private final String stepName;
+
+  private MetricsMap<MetricName, CounterCell> counters =
+      new MetricsMap<MetricName, CounterCell>() {
+        @Override
+        protected CounterCell createInstance() {
+          return new CounterCell();
+        }
+      };
+
+  private MetricsMap<MetricName, DistributionCell> distributions =
+      new MetricsMap<MetricName, DistributionCell>() {
+        @Override
+        protected DistributionCell createInstance() {
+          return new DistributionCell();
+        }
+      };
+
+  /**
+   * Create a new {@link MetricsContainer} associated with the given {@code stepName}.
+   */
+  public MetricsContainer(String stepName) {
+    this.stepName = stepName;
+  }
+
+  /**
+   * Set the {@link MetricsContainer} for the current thread.
+   */
+  public static void setMetricsContainer(MetricsContainer container) {
+    CONTAINER_FOR_THREAD.set(container);
+  }
+
+  /**
+   * Clear the {@link MetricsContainer} for the current thread.
+   */
+  public static void unsetMetricsContainer() {
+    CONTAINER_FOR_THREAD.remove();
+  }
+
+  /**
+   * Return the {@link MetricsContainer} for the current thread.
+   */
+  public static MetricsContainer getCurrentContainer() {
+    MetricsContainer container = CONTAINER_FOR_THREAD.get();
+    if (container == null) {
+      throw new IllegalStateException("Must call setMetricsContainer before reporting metrics.");
+    }
+    return container;
+  }
+
+  /**
+   * Return the {@link CounterCell} that should be used for implementing the given
+   * {@code metricName} in this container.
+   */
+  public CounterCell getOrCreateCounter(MetricName metricName) {
+    return counters.getOrCreate(metricName);
```

Comment:
I'm not completely sure what you mean -- I suspect you're proposing either inlining the logic from MetricsMap#getOrCreate into each of these methods or passing a function to MetricsMap#getOrCreate to do that.

I actually prefer leaving it like it is, although I'm open to renaming the class. There are several different places where we need the functionality of a "create-on-demand, concurrent map" and it is useful to have that in one place so we can make changes in that one place.